Format error message for unmatched signature between _out and base functions

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47087 Format error message for unmatched signature between _out and base functions**

Updated Error Message:
![image](https://user-images.githubusercontent.com/68879799/97923622-46e34c80-1d2c-11eb-8d74-84f8649064be.png)

Fixes #33547

Differential Revision: [D24633077](https://our.internmc.facebook.com/intern/diff/D24633077)

ejguan committed:
Format error message for unmatched signature between _out and base functions

[ghstack-poisoned]

ejguan committed:
Update on "Format error message for unmatched signature between _out and base functions"


Fixes #33547


[ghstack-poisoned]

zou3519 commented on tools/autograd/gen_python_functions.py:
```diff
@@ -702,8 +702,8 @@ def group_overloads(declarations, is_python_method):
             raise RuntimeError(
-                "'base' not in dictionary for {}. keys are {}".format(
-                    x, list(dictionary.keys())))
+                "_out variant function {} can not find matched non-out 'base' function "
+                "{} in dictionary.".format(dictionary['signature'], x))
```
Let's try to make this more human readable:
```
While identifying overloads, we found an out= schema {out_schema} without a corresponding non-out variant.
We expected the non-out variant to have schema {expected_schema}. Please check that you spelled the schema correctly in native_functions.yaml.
```

I think the `declarations` argument to gen_python_functions might have possible candidates. If so, we can make the error message even more prescriptive:

```
While identifying overloads, we found an out= schema {out_schema} without a corresponding non-out variant.
We expected the non-out variant to have schema {expected_schema}. Please check that you spelled the schema correctly in native_functions.yaml. We discovered the following candidates:
- {candidate1}
- {candidate2}
...
```

ejguan replied:
> I think the `declarations` argument to gen_python_functions might have possible candidates. If so, we can make the error message even more prescriptive:

Yes, it's doable. And the overhead for a whole loop through `declarations` to find the non-out function is minimum.

ejguan commented:
Update error message

ejguan committed:
Update on "Format error message for unmatched signature between _out and base functions"


Current Error Message:
![image](https://user-images.githubusercontent.com/68879799/97733320-12faf380-1aae-11eb-80b3-bf2d336403e2.png)

Fixes #33547

Differential Revision: [D24633077](https://our.internmc.facebook.com/intern/diff/D24633077)

[ghstack-poisoned]

zou3519 commented on tools/autograd/gen_python_functions.py:
```diff
@@ -701,9 +701,17 @@ def group_overloads(declarations, is_python_method):
         if 'base' not in dictionary:
+            candidates = []
+            non_out_name = dictionary['out']['operator_name']
+            for declaration in declarations:
+                if declaration['name'] == non_out_name and not declaration['deprecated']:
+                    signature = get_python_signature(declaration, is_python_method, skip_outputs=True)
+                    candidates.append(signature)
             raise RuntimeError(
-                "'base' not in dictionary for {}. keys are {}".format(
-                    x, list(dictionary.keys())))
+                "While identifying overloads, we found an out schema {} without a corresponding non-out variant. "
+                "We expected the non-out variant to have schema: \n- {}\nPlease check that you spelled the schema "
+                "correctly or in native_functions.yaml. We discovered the following candidate(s): \n"
+                .format(dictionary['signature'], x) + "\n".join("- {}".format(candidate) for candidate in candidates))
```
Can you update the sample error message in the PR body?

ejguan replied:
Sure. There is actually a typo in the error message, and I will fix it and update PR.

ejguan committed:
Update on "Format error message for unmatched signature between _out and base functions"


Updated Error Message:
![image](https://user-images.githubusercontent.com/68879799/97923622-46e34c80-1d2c-11eb-8d74-84f8649064be.png)

Fixes #33547

Differential Revision: [D24633077](https://our.internmc.facebook.com/intern/diff/D24633077)

[ghstack-poisoned]
